File: scala/src/main/scala/coffee/lambda/Ex04ListCombinatorsSolution.scala
package coffee.lambda

/* Using only the library foldLeft and foldRight methods (no primitive recursion, iteration etc.)
 * implement the following list combinators:
 * a) reverse
 * b) map
 * c) filter
 * d) append
 * e) concat
 */

object Ex04ListCombinatorsSolution {
  def reverse[T](list: List[T]): List[T] = list.foldLeft(List[T]())((list, elt) => elt :: list)

  def map[T, R](list: List[T], fun: T => R): List[R] = list.foldRight(List[R]())((elt, acc) => fun(elt) :: acc)

  def filter[T](list: List[T], pred: T => Boolean): List[T] = list.foldRight(List[T]())(
    (elt, acc) => if (pred(elt)) elt :: acc else acc
  )

  def append[T](a: List[T], b: List[T]): List[T] = a.foldRight(b)(_ :: _)

  def concat[T](a: List[List[T]]): List[T] = a.foldRight(List[T]())(append)
}

File: haskell/ex04_list_combinators_solution.hs
import Test.HUnit

{-
 - Problem 4: List Combinators.
 - Using only foldl and foldr (no recursion etc.) implement the following list combinators:
 - a) reverse'
 - b) map'
 - c) filter'
 - d) concat'
 - Bonus points for using currying and point-free notation :).
 -}

reverse' :: [a] -> [a]
reverse' = foldl (flip (:)) []

map' :: (a -> b) -> [a] -> [b]
map' f = foldr (\el acc -> (f el) : acc) []

filter' :: (a -> Bool) -> [a] -> [a]
filter' f = foldr (\el acc -> if f el then el : acc else acc) []

concat' :: [[a]] -> [a]
concat' = foldr (flip $ foldr (:)) []

--------------------------------------------------------------------------------
-----------------------------------TEST CASES-----------------------------------
--------------------------------------------------------------------------------

testReverse = TestCase $ assertEqual "revese' reverses" [3, 2, 1] (reverse' [1, 2, 3])

testMap = TestCase $ assertEqual "map' maps" [2, 3, 4] (map' (+ 1) [1, 2, 3])

testFilter = TestCase $ assertEqual "filter' filters" [2, 4] (filter' even [1, 2, 3, 4])

testConcat = TestCase $ assertEqual "concat' concats" [1, 2, 3, 4, 5] (concat' [[1, 2], [3, 4], [5]])

main = runTestTT $ TestList [ testReverse
                            , testMap
                            , testFilter
                            , testConcat ]
